hopsworks-api/src/main/java/io/hops/hopsworks/api/metadata/XAttrsResource.java

```diff
@@ -0,0 +1,169 @@
+import io.hops.hopsworks.exceptions.DatasetException;
+import io.hops.hopsworks.exceptions.MetadataException;
+import io.hops.hopsworks.jwt.annotation.JWTRequired;
+import io.hops.hopsworks.restutils.RESTCodes;
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiOperation;
+import org.json.JSONObject;
+
+import javax.ejb.EJB;
+import javax.ejb.TransactionAttribute;
+import javax.ejb.TransactionAttributeType;
+import javax.enterprise.context.RequestScoped;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.SecurityContext;
+import javax.ws.rs.core.UriBuilder;
+import javax.ws.rs.core.UriInfo;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+
+@Api(value = "Extended Attributes Resource")
+@RequestScoped
+@TransactionAttribute(TransactionAttributeType.NEVER)
+public class XAttrsResource {
+  
+  @EJB
+  private ProjectFacade projectFacade;
+  @EJB
+  private JWTHelper jWTHelper;
+  @EJB
+  private XAttrsController xattrsController;
+  
+  private Project project;
+  
+  public void setProject(Integer projectId) {
+    this.project = projectFacade.find(projectId);
+  }
+  
+  
+  @ApiOperation( value = "Create or Update an extended attribute for a path.")
+  @PUT
+  @Path("{path: .+}")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  @AllowedProjectRoles({AllowedProjectRoles.DATA_SCIENTIST, AllowedProjectRoles.DATA_OWNER})
+  @JWTRequired(acceptedTokens={Audience.API}, allowedUserRoles={"HOPS_ADMIN", "HOPS_USER"})
+  public Response put(@Context
+      SecurityContext sc, @Context
+      UriInfo uriInfo, @PathParam("path") String path,
+      @QueryParam("name") String xattrName,
+      String metaObj) throws DatasetException,
+      MetadataException {
+    Users user = jWTHelper.getUserPrincipal(sc);
+    
+    if(xattrName == null || xattrName.isEmpty())
+      return Response.noContent().status(Response.Status.BAD_REQUEST).build();
+    
+    if (metaObj.length() > 13500 || xattrName.length() > 255) {
+      throw new MetadataException(RESTCodes.MetadataErrorCode.METADATA_MAX_SIZE_EXCEEDED, Level.FINE);
+    }
+    Response.Status status = Response.Status.CREATED;
+    String metadata = xattrsController.getXAttr(project, user, path,
+        xattrName);
+    if(metadata != null){
+      status = Response.Status.OK;
+    }
+    
+    xattrsController.addXAttr(project, user, path, xattrName,
+        metaObj);
+    
+    UriBuilder builder = uriInfo.getAbsolutePathBuilder();
+    if(status == Response.Status.CREATED) {
+      return Response.created(builder.build()).build();
+    } else {
+      return Response.ok(builder.build()).build();
```

Comment:
@maismail PUT endpoint should return the created entity https://github.com/logicalclocks/hopsworks/blob/v1.0.0/hopsworks-api/src/main/java/io/hops/hopsworks/api/jobs/JobsResource.java#L196

Reply:
Done.